### PR TITLE
Healing the tree of life - fixing _swap_subtrees_move_ for same lineage moves. 

### LINKED ISSUE #47: Consider cleaning up `tree_utils`
In `tree_inference.tree_utils` there are 3 functions from building the tree moves that are not used. Only used for testing purposes. 

`_prune`, `_reattach`, `prune_and_reattach_move`

One could consider moving these to tests, to keep the main code short.

### LINKED ISSUE #48: Investigate Potential Bug in MCMC Tree Moves
During the visualization issue #42 some trees in an outdated, hence now irreproducible dataset, outputted by the MCMC sampler were not convertible to the `TreeNode` format. The reason was, that they seemed to lack a root. - Or the last node in the adjacency matrix at least was not the root. The exact issue is not clear. The function `tree_inference_tree._get_root_label` got an empty array in their line 
`# find row which has all ones in ancestor_matrix
    root_idx = jnp.where(jnp.all(ancestor_matrix == 1, axis=1))[0]`
    
- [x] Consider adding debug logging to the tree moved to discover what went wrong the next time this happens.
- [x] Add tests for the individual tree moves and check if it is a valid tree
- [x] identify the issue
- [x] fix the issue
- [x] rerun mcmc mark00
- [x] check if all trees can be viszualized
- [x] check speed of mcmc

Conjectures:
- root is assumed to be the last in the tree, and was swapped to be not the last
- confusion if to include the root in moves or not

### PR DESCRIPTION
**Pull request by gordonkoehn**
This PR addresses/closes #48 mainly, it also closes #47. 

The `tree_inference.mcmc._swap_subtrees_move` and `tree_inference.mcmc._swap_subtrees_proposal` were incomplete/wrong for the case of choosing nodes of the same lineage. Now fixed and tested against the SCITE supplement paper examples. 

The failure of this tree move became apparent during the coding of the _Visualization of the Timeseries_ in #42, as some trees were broken into two and mcmc samples could not be imported.

To convince myself the tree moves now make sense and the trees don't break, I decided to include a small visualization script of the trees of a mcmc chain in this PR. This resulted in a very satisfying animation. See here: 

animation

These are the main objectives of this PR. 

It took a bit of tracing of the issues, hence note the addition of `loggers` in many scripts. 

Merging this PR will allow us to proceed with the  _Visualization of the Timeseries_ in #42.

Apologies for the large and messy PR - the result of some digging for the bug and the self-convincing of its workings.

### CODE REVIEW

File: src/pyggdrasil/visualize/tree.py [hunk 1]
```diff
@@ -71,8 +81,22 @@ def plot(
...
+    node_height = 0.75  # Height of each node in the figure
+    figure_width = width * node_width
+    figure_height = depth * node_height
+
+    # print(f"figure_width: {figure_width}")
+    # print(f"figure_height: {figure_height}")
+
     # plot
-    fig = plt.figure()
+    fig = plt.figure(figsize=(figure_width, figure_height))
```
Comment (gordonkoehn) [inline src/pyggdrasil/visualize/tree.py:99]: Dynamic figure sizing depending on tree depth and breadth! Rather proud of this chatGPT-assisted gimmick. Works well with our 50-node trees.

File: tests/tree_inference/test_mcmc.py [hunk 2]
```diff
@@ -191,3 +193,161 @@ def test_swap_subtrees_move():
...
+
+    # check if any of the four possible trees is the same as the new tree
+    assert (
+        jnp.array_equal(new_tree.tree_topology, new_tree_corr1.tree_topology)
+        or jnp.array_equal(new_tree.tree_topology, new_tree_corr2.tree_topology)
+        or jnp.array_equal(new_tree.tree_topology, new_tree_corr3.tree_topology)
+        or jnp.array_equal(new_tree.tree_topology, new_tree_corr4.tree_topology)
+    )
+    # check if the labels are the same, just once for sanity
+    assert jnp.array_equal(new_tree.labels, new_tree_corr1.labels)
```
Comment (gordonkoehn) [inline tests/tree_inference/test_mcmc.py:353]: Implements the different lineage subtree swap as in the SCITE supplement. 

Screenshot 2023-05-30 at 13 02 38

File: tests/visualize/test_viz_tree.py [hunk 3]
```diff
@@ -9,7 +9,7 @@
 # from pathlib import Path
 
 from pyggdrasil.tree_inference._tree import Tree
-import pyggdrasil.visualize.tree as viz
+import pyggdrasil.visualize as viz
```
Comment (gordonkoehn) [inline tests/visualize/test_viz_tree.py:12]: fixing public API

**Review by gordonkoehn — COMMENTED** [review]
Ready for review - refactorings may distract from the main changed. Main changes are marked.

File: pyproject.toml [hunk 4]
```diff
@@ -22,6 +22,7 @@ datetime = "^5.1"
 pytz = "^2023.3"
 pydantic = "^1.10.7"
 cyvcf2 = "<0.30.20"  # Temporary dependency: see https://github.com/cbg-ethz/PYggdrasil/pull/40/
+tqdm = "^4.65.0"
```
[the remark this answers is not in the document]
Comment (pawel-czyz) [reply pyproject.toml:25]: Yes, also tqdm has a `disable` argument, which can hide it.

File: scripts/plot_trees.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    Returns:
+        None
+    """
+
+    args = create_parser()
+
+    # get the full path
+    out_dir = Path(args.out_dir)
+    # if out_dir does not exist, create it
+    if not out_dir.exists():
```
Comment (pawel-czyz) [inline scripts/plot_trees.py:70]: I don't really get why the `if` statement is needed here – wouldn't `exist_ok=False` work here?

File: scripts/plot_trees.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        out_dir.mkdir(parents=True, exist_ok=True)
+
+    # Set up logging
+    fullpath_log = out_dir / "plot_trees.log"
```
Comment (pawel-czyz) [inline scripts/plot_trees.py:74]: If `out_dir` is a `Path`, then `fullpath_log` is also a `Path` and manual casting below is not really needed :wink:

File: scripts/plot_trees.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    if not Path(fullpath_log).exists():
+        Path(fullpath_log).touch()
+    # if logfile exist delete it and create a new one
+    else:
+        Path(fullpath_log).unlink()
+        Path(fullpath_log).touch()
+
+    logging.basicConfig(filename=fullpath_log, level=logging.INFO)
+    # set logging level for jax
+    logging.getLogger("jax._src.dispatch").setLevel(logging.ERROR)
```
Comment (pawel-czyz) [inline scripts/plot_trees.py:85]: I see what this is for, nice!

File: scripts/plot_trees.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+    # print options
+    print_options = dict()
+    print_options["title"] = False
+    print_options["data_tree"] = dict()
+    print_options["data_tree"]["log-likelihood"] = True
+
+    # for each iteration, plot the tree
+    for i in tqdm(iterations):
+        i = int(i)
```
Comment (pawel-czyz) [inline scripts/plot_trees.py:109]: I don't think this `int(i)` is really needed, is it? `tqdm` should return just the original object.

File: src/pyggdrasil/interface/__init__.py [hunk 9]
```diff
@@ -47,10 +56,13 @@ def get_sample(self, iteration: int) -> tuple[int, TreeNode, float]:
...
+        # get index of iteration
+        iteration_idx = jnp.where(self.iterations == iteration)[0][0]
+
         return (
             iteration,
-            self.trees[iteration],
-            self.log_probabilities[iteration].item(),
+            self.trees[iteration_idx],
+            self.log_probabilities[iteration_idx].item(),
         )
```
[the remark this answers is not in the document]
Comment (pawel-czyz) [reply src/pyggdrasil/interface/__init__.py:66]: This is subtle and I'd really like to understand this: why `iteration_idx` can be different from `iteration`? Is it off by one error?

```
iteration_idx = jnp.where(self.iterations == iteration)[0][0]
```

File: src/pyggdrasil/tree_inference/_analyze.py [hunk 10]
```diff
@@ -2,13 +2,18 @@
...
 from pyggdrasil import TreeNode, compare_trees
 from pyggdrasil.tree_inference import Tree, unpack_sample
 
 from pyggdrasil.interface import MCMCSample, PureMcmcData
 
 
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+
+
```
[the remark this answers is not in the document]
Comment (pawel-czyz) [reply src/pyggdrasil/tree_inference/_analyze.py:16]: Very good idea :slightly_smiling_face: Indeed, if each module has a separate logger it's easier to see what the program is doing.

[the remark this answers is not in the document]
Comment (pawel-czyz) [reply src/pyggdrasil/tree_inference/_analyze.py:16]: By the way, you may like reading about https://www.structlog.org/en/stable/ (even if it doesn't get used here – I used it once, when I was working on a large pipeline and automated analysis of any failures/caught exceptions/whatever unexpected happened was essential).

File: src/pyggdrasil/tree_inference/_mcmc.py [hunk 11]
```diff
@@ -39,6 +42,11 @@ def _prune_and_reattach_move(tree: Tree, *, pruned_node: int, attach_to: int) ->
     # make new tree
     new_tree = Tree(tree_topology=new_adj_mat, labels=tree.labels)
 
+    logger.debug(
+        "MCMC: Prune and reattach move - pruned node %s; attached to node %s",
+        pruned_node,
+        attach_to,
+    )
```
[the remark this answers is not in the document]
Comment (pawel-czyz) [reply src/pyggdrasil/tree_inference/_mcmc.py:49]: I don't think it impacts the speed by any significant factor.
I think it's wonderful that you added them and traced the bug, so unless we need to optimise we can leave them here. (And if we want to optimise, we'll go to the fully JAX-compliant implementation anyway, I think.)

File: src/pyggdrasil/tree_inference/_mcmc.py [hunk 12]
```diff
@@ -124,15 +138,23 @@ def _swap_node_labels_proposal(
...
-def _swap_subtrees_move(tree: Tree, node1: int, node2: int) -> Tree:
-    """Swaps subtrees rooted at ``node1`` and ``node2``.
+def _swap_subtrees_move(
+    tree: Tree,
+    node1: int,
+    node2: int,
+    same_lineage: bool,
+    key: random.PRNGKeyArray = None,
+) -> Tree:
+    """Swaps subtrees rooted at ``node1``/ ``node i`` and ``node2``/ ``node k``.
```
[the remark this answers is not in the document]
Comment (pawel-czyz) [reply src/pyggdrasil/tree_inference/_mcmc.py:148]: This answer is going to be very subjective and opinionated, but:
  - in Python overloading is very limited, although sometimes useful for type checking – see e.g., [this tutorial](https://adamj.eu/tech/2021/05/29/python-type-hints-how-to-use-overload/);
  - whenever different sets of arguments are passed, it is a potential sign that *this function does two things and should be refactored into two functions*. (Recall [SRP](https://en.wikipedia.org/wiki/Single-responsibility_principle)).

I think this is a good example: if the case of being in the same lineage is so important, it should perfectly be in a different function. It will simplify the logic.

Then, the move (which has access to the random key) can decide which of these functions should be used (by comparing the nodes and observing whether they are in the same lineage).

I'm not sure if this writing is clear and coherent enough – let me know if not, we can also discuss it in the meeting :slightly_smiling_face:

File: src/pyggdrasil/tree_inference/_mcmc.py [hunk 13]
```diff
@@ -145,14 +167,48 @@ def _swap_subtrees_move(tree: Tree, node1: int, node2: int) -> Tree:
...
+        # make new tree
+        new_tree = Tree(tree_topology=new_adj_mat, labels=tree.labels)
+
+        logger.debug(
+            "MCMC: Swap subtrees move - nested swap of"
+            " subtrees rooted at node %s and node %s",
+            node1,
+            node2,
+        )
+        return new_tree
```
[the remark this answers is not in the document]
Comment (pawel-czyz) [reply src/pyggdrasil/tree_inference/_mcmc.py:211]: Great you've found it!

I should've actually caught that in the review implementing the move. I need to improve on my false negative rate :sweat_smile:

File: src/pyggdrasil/visualize/tree.py [hunk 14]
```diff
@@ -55,6 +59,12 @@ def plot(
     Note:
         see tests/visualize/test_tree.py for example usage
     """
+
+    # check if tree-name is none and throw log - and error
+    if tree.name is None and print_options["title"] is True:
+        logger.error("Tree name is None, cannot print title")
+        raise ValueError("Tree name is None, cannot print title")
+
```
[the remark this answers is not in the document]
Comment (pawel-czyz) [reply src/pyggdrasil/visualize/tree.py:67]: I see :slightly_smiling_face:

I'm not sure if it'd be cleaner in this case, but sometimes using a non-union type may be also an option (e.g., `name` could always be a `str` or set by default to `""`).

File: src/pyggdrasil/visualize/tree.py (hunk 1 above)
Comment (pawel-czyz) [reply src/pyggdrasil/visualize/tree.py:99]: Oh, this is very cool! Now I see how you generated this video! :slightly_smiling_face: 


@monicadragan I know that you have a whole JS library for drawing trees, but perhaps it'd be of interest also to you :slightly_smiling_face:

File: tests/tree_inference/test_mcmc.py (hunk 2 above)
Comment (pawel-czyz) [reply tests/tree_inference/test_mcmc.py:353]: Great job! :slightly_smiling_face:

**Review by pawel-czyz — APPROVED** [review]
This is *really good*. I think it's great you found this bug and as a side effect you added logging and visualisation utilities.

My only real comment (and which doesn't need to be addressed before merging) is that the swap subtree move should really be based on two functions: one for nodes in the same lineage and the other one. They take different arguments and have different logic, what can be seen in the long `if`/`else` block.

But the current version works and you added a nice test from the paper – looks as ready to merge! Great work :slightly_smiling_face:

File: src/pyggdrasil/tree_inference/_mcmc.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (gordonkoehn) [reply src/pyggdrasil/tree_inference/_mcmc.py:49]: Great, would be excited to see how the optimization with Jax would work. :)

File: src/pyggdrasil/visualize/tree.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (gordonkoehn) [reply src/pyggdrasil/visualize/tree.py:67]: Yes, I see. - I think I'll keep it this way, just to be explicit.

File: src/pyggdrasil/visualize/tree.py (hunk 1 above)
Comment (gordonkoehn) [reply src/pyggdrasil/visualize/tree.py:99]: Still experimental - not yet tested on all sensible tree sizes.

File: scripts/plot_trees.py (hunk 5 above)
Comment (gordonkoehn) [reply scripts/plot_trees.py:77]: Good point! Made a few changes to be more efficient here.

File: scripts/plot_trees.py (hunk 6 above)
Comment (gordonkoehn) [reply scripts/plot_trees.py:80]: Oh great. - overlooked this.

File: scripts/plot_trees.py (hunk 7 above)
Comment (gordonkoehn) [reply scripts/plot_trees.py:87]: Only really needed if one sets the root logging to `DEBUG`. But I left it, for convenience for future bugs.

File: src/pyggdrasil/tree_inference/_mcmc.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (gordonkoehn) [reply src/pyggdrasil/tree_inference/_mcmc.py]: Very interesting! As always, I greatly apparate these nuggets of new input :) 

Tried to comply with SRP now. Split the function up into 3. 
1) `detrach the subtrees` 2) same lineage 3) different lineage

But still need to keep the choice of same or different lineage in the `swap_subtrees_proposal`, needed for the correction factor.

**Review by gordonkoehn — COMMENTED** [review]
all good now. -  all feedback in.
